chore: 1031 migrate docs to nuxt UI docs template (#1049)
Tresjs/tres: docs/content/3.api/3.advanced/performance.md

---
title: Scaling Performance 🚀
description: Quick guide with tips to improve the performance of your TresJS application.
---

Running WebGL in the browser can be resource-intensive depending on the user's device capabilities. To make 3D accessible to everyone, it's important to optimize your applications for performance, especially on low-end devices. This guide shares practical tips to help you get the best performance from your TresJS projects.

> ## On-demand Rendering

Author: This one appears duplicated I think, I addressed it twice

<script setup lang="ts">
import { reactive } from 'vue'

// This will make all properties reactive (not recommended for Three.js objects)
const position = reactive({ x: 0, y: 0, z: 0 })

onLoop(({ _delta, elapsed }) => {
  position.x = Math.sin(elapsed * 0.1) * 3
})
</script>

<template>
  <TresMesh :position="position" cast-shadow>
    <TresBoxGeometry :args="[1, 1, 1]" />
    <TresMeshStandardMaterial color="teal" />
  </TresMesh>
</template>
```

### ✅ Correct: Use shallowRef and Direct Assignment

```vue [correct-reactivity.vue]
<script setup lang="ts">
import { shallowRef } from 'vue'

const position = { x: 0, y: 0, z: 0 }
const boxRef = shallowRef(null)

onLoop(({ _delta, elapsed }) => {
  if (boxRef.value) {
    boxRef.value.position.x = Math.sin(elapsed * 0.1) * 3
  }
})
</script>

<template>
  <TresMesh ref="boxRef" :position="position" cast-shadow>
    <TresBoxGeometry :args="[1, 1, 1]" />
    <TresMeshStandardMaterial color="teal" />
  </TresMesh>
</template>
```

::note
**Tip:** Only the `.value` of a `shallowRef` is reactive. The internal properties are not, which is ideal for Three.js objects that are updated frequently.
::

## Dispose Resources with `dispose()`

> When you're finished with a resource—such as a texture, geometry, or material—be sure to dispose of it to free up memory. This is especially important if your app frequently creates and destroys resources, like in games or interactive experiences.

Reviewer: ```suggestion
When a resource is no longer needed—such as a texture, geometry, or material—be sure to dispose of it to free up memory. This is especially important if your app frequently creates and destroys resources, like in games or interactive experiences.
```
